Added new getting started guide for node js

liam-b opened the pull request.

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+---
+
+{% include /style/icon.html type="warning" %}
+Note that Node JS run really slowly on the EV3 brick and is limited to NPM version 0.10.29 which can be a problem when trying to install some libraries.
```
`run` -> `runs`

WasabiFan replied:
"really" sounds mildly unprofessional in this context. Can we use another word there?

WasabiFan replied:
`NPM` is `npm`

WasabiFan replied:
It's also limited to an equivalently old Node.js version, which should also be mentioned.

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+Note that Node JS run really slowly on the EV3 brick and is limited to NPM version 0.10.29 which can be a problem when trying to install some libraries.
+{: .alert .alert-warning}
+
+**Before you start, make sure that you have configured a network connection to
```
Can you link to the tutorials that we have for these tasks?

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+{% include /style/icon.html type="warning" %}
+Note that Node JS run really slowly on the EV3 brick and is limited to NPM version 0.10.29 which can be a problem when trying to install some libraries.
+{: .alert .alert-warning}
+
```
It might make this page more approachable to have a one-sentence explanation of what Node.js is and what it's used for. You can format it as a ["lead" paragraph](https://github.com/ev3dev/ev3dev.github.io/blob/master/projects/index.html#L7) if you'd like.

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+your ev3dev device and have opened an SSH connection to it.**
+
+## Installing ev3dev-lang
+To install the library run:
```
Without an explicit mention that this should be run via SSH, some people will try running this on their PC :laughing:

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
````diff
@@ -0,0 +1,100 @@
+var ev3dev = require('ev3dev-lang');
+```
+
+For a bit more info on this visit the ev3dev-lang [github page](https://github.com/wasabifan/ev3dev-lang-js).
````
`github` -> `GitHub`

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+For this tutorial, we will be using nano. For more info on how to use nano check out the [nano cheat sheet](/docs/tutorials/nano-cheat-sheet).
+
+## Hello world 
+Navigate to a project directory (e.g `~/src/js`) and make a new file with `touch test.js`
```
`e.g` -> `e.g.`

WasabiFan replied:
Period after sentence

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
````diff
@@ -0,0 +1,100 @@
+console.log(test);
+```
+
+Running `node test.js`should print `hello world!` to the console. If this worked you are ready to start coding with node!
````
Although I've often formatted the name as "node" or "Node," they [officially use "Node.js"](https://nodejs.org/en/); I think we should align with that wherever possible.

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+Running `node test.js`should print `hello world!` to the console. If this worked you are ready to start coding with node!
+
+## Motors
+Plug a ev3 large motor into port A, then open `test.js` and write:
```
`a ev3` -> `an EV3`

WasabiFan replied:
People might need to be reminded how to open files again. Maybe you can just link to the cheat sheet everywhere that you mention nano :laughing: Your choice.

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+Plug a ev3 large motor into port A, then open `test.js` and write:
+
+{% include /style/icon.html type="warning" %}
+This snippet of code will not work for any other motors!
```
Mention what other motors the user might have (so they know how to tell them apart)

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+
+var motor = new ev3dev.LargeMotor('outA'); // create new motor plugged into port A called 'motor'
+
+motor.runForever(200); // run motor at speed 200
```
I'd prefer to not tell people that this is even an option for running for a length of time, given that there is a `runForTime` method. Do you have any other ideas for a good demo?

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+
+If you get any errors when you try to run this, make sure your motor is plugged in correctly.
+
+For other motors such as the `MeduimMotor`, check the [default supported motors](http://wasabifan.github.io/ev3dev-lang-js/modules/_motors_.html) list.
```
`Meduim` -> `Medium`

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+
+For other motors such as the `MeduimMotor`, check the [default supported motors](http://wasabifan.github.io/ev3dev-lang-js/modules/_motors_.html) list.
+
+And from reading the [documentation page](http://wasabifan.github.io/ev3dev-lang-js) you should be able to get a good idea of what methods you can use with the motor (things like `runForever()` and `runForTime()`). Then when you use it in code, it should look like:
```
Remove "And" at the beginning of the sentence

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
````diff
@@ -0,0 +1,100 @@
+motor.doSomething(speed, otherArgument) // where doSomething can be replaced by any valid method.
+```
+
+You can also **reset** and **read** the position of a motor using `motor.reset()` and `motor.position`.
````
Keep in mind that `reset` resets the position _and the last-used command, stop mode, etc._ Resetting the position alone is `motor.position = 0`. To make this clear, you could do this:

> You can also **reset** and **read the position of** a motor using `motor.reset()` and `motor.position`, respectively.

What you do exactly depends on what you'd like to communicate there; I'll leave that up to you.

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+You can also **reset** and **read** the position of a motor using `motor.reset()` and `motor.position`.
+
+## Sensors
+Plug a ev3 color sensor into port 1, then edit `test.js` again appending this to the bottom:
```
`EV3`

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+
+For other sensors, check the [default supported sensors](http://wasabifan.github.io/ev3dev-lang-js/modules/_sensors_.html) list.
+
+To change a sensors mode you'll need to check what modes it has by looking at the [ev3dev sensor list](http://docs.ev3dev.org/projects/lego-linux-drivers/en/ev3dev-jessie/sensor_data.html) and use:
```
`sensors` -> `sensor's`

WasabiFan replied:
What do you think people will need to set the mode for? Properties like `reflectedLightIntensity` should automatically set it.

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
````diff
@@ -0,0 +1,100 @@
+sensor.mode = 'NEW-MODE'; // where 'NEW-MODE' can be replaced by any valid mode
+```
+
+And then using `sensor.someValueAcessor` will return the sensor value, where `someValueAcessor` can be replaced by any valid accessor (also found on the [default supported sensors](http://wasabifan.github.io/ev3dev-lang-js/modules/_sensors_.html) list by clicking on the sensor you are using)
````
`accessor` (two Cs)

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+And then using `sensor.someValueAcessor` will return the sensor value, where `someValueAcessor` can be replaced by any valid accessor (also found on the [default supported sensors](http://wasabifan.github.io/ev3dev-lang-js/modules/_sensors_.html) list by clicking on the sensor you are using)
+
+## Next steps
+So that's the basic overview of how to use node with ev3dev! For info on all supported sensors, motors and other documentation visit the [github page](https://github.com/wasabifan/ev3dev-lang-js) and the [documentation page](http://wasabifan.github.io/ev3dev-lang-js).
```
Remove "So"

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+## Next steps
+So that's the basic overview of how to use node with ev3dev! For info on all supported sensors, motors and other documentation visit the [github page](https://github.com/wasabifan/ev3dev-lang-js) and the [documentation page](http://wasabifan.github.io/ev3dev-lang-js).
+
+The documentation page looks a bit confusing, but what you really need to worry about are the names of the classes and the **methods** / **accessors**. (eg. the large motor has **methods** such as: `runForever()`, `runForTime()` etc and the color sensor has **accessors** such as: `reflectedLightIntensity`, `ambientLightIntensity` etc)
```
> `ambientLightIntensity` etc

should be

> `ambientLightIntensity`, etc.

WasabiFan commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,100 @@
+
+The documentation page looks a bit confusing, but what you really need to worry about are the names of the classes and the **methods** / **accessors**. (eg. the large motor has **methods** such as: `runForever()`, `runForTime()` etc and the color sensor has **accessors** such as: `reflectedLightIntensity`, `ambientLightIntensity` etc)
+
+The [sensor list](http://docs.ev3dev.org/projects/lego-linux-drivers/en/ev3dev-jessie/sensor_data.html) and the [motor list](http://docs.ev3dev.org/projects/lego-linux-drivers/en/ev3dev-jessie/motor_data.html) are both really useful pages aswell.
```
`really` -> `extremely`

WasabiFan replied:
`aswell` -> `as well`

WasabiFan commented:
Sorry about the Travis CI server error; it's currently having some issues. Your page doesn't have any errors.

liam-b commented:
Hey! Yeah, I'm not the greatest at writing documentation :). Thanks for editing it though! I added 'getting started guides' to the description of that category and went through and fixed what you pointed out.

Just a quick question, what other versions are npm limited to? Because it's 0.10.29 on my brick.

Thanks,
Liam

WasabiFan commented:
> Yeah, I'm not the greatest at writing documentation :).

No problem! It's always easier to spot issues as a reviewer than as the original author... You grow blind out of familiarity.

> Just a quick question, what other versions are npm limited to? Because it's 0.10.29 on my brick.

npm is at that version; my comment was that Node.js itself is stuck on 0.10 or 0.12 (I can't remember exactly) and support has essentially been dropped for the EV3's processor since.

WasabiFan commented:
I added my suggested changes to this PR; @liam-b if you disagree with any of them, feel free to edit it back or ask!

@dlech can you take a quick pass over this?

dlech commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -5,74 +5,70 @@ author: "@liam-b"
 
 {% include /style/icon.html type="warning" %}
-Note that Node JS run really slowly on the EV3 brick and is limited to NPM version 0.10.29 which can be a problem when trying to install some libraries.
+Note that Node.js runs frustratingly slowly on the EV3 brick and is limited to an old version of both Node.js and npm. You may want to consider using Python or another ev3dev-supported language instead.
```
It looks like non-EV3 stretch images are going to have node 4.7.2

dlech commented on _data/tutorial-groups.yml:
```diff
@@ -9,7 +9,7 @@
   - id: software-languages
     title: Programming Languages and Tooling
-    description: "Configuring a development environment, installing new languages, and using advanced language tools."
+    description: "Configuring a development environment, installing new languages, using advanced language tools and getting started guides for languages."
```
It might be better to put "getting started" first in the list.

dlech commented on docs/tutorials/getting-started-with-node-js.md:
```diff
@@ -0,0 +1,92 @@
+
+<p class="lead">
+Node.js is a framework that allows you to run JavaScript as a local application. It is designed to make building servers and other asynchronous apps as easy as possible. In this tutorial, we'll show you how to get up-and-running with Node.js on ev3dev.
+</p>
```
Just do `{: .lead}` at the end here. Kramdown automatically creates the paragraph element.

dlech commented on docs/tutorials/getting-started-with-node-js.md:
````diff
@@ -0,0 +1,92 @@
+To start, navigate to a project directory (e.g `~/src/js`) that we can store our files in:
+
+```shell
+mkdir -p ~/src/js && cd ~/src/js
````
This might be better as 2 separate lines instead of `&&`

dlech commented on docs/tutorials/getting-started-with-node-js.md:
````diff
@@ -0,0 +1,92 @@
+```javascript
+var ev3dev = require('ev3dev-lang');
+
+var motor = new ev3dev.Motor('outA'); // create new motor plugged into port A called 'motor'
````
Might want a link to <http://www.ev3dev.org/docs/platform-comparison/> here to get the port names for non-EV3 hardware.
